### PR TITLE
Install pip metadata

### PR DESCRIPTION
**Pull request by traversaro**
Fix https://github.com/conda-forge/opencv-feedstock/issues/299 .

Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by traversaro** [comment]
@conda-forge-admin, please rerender

File: recipe/meta.yaml [hunk 1]
```diff
@@ -253,6 +254,9 @@ outputs:
         - if [[ $($PYTHON -c 'import cv2; print(cv2.__version__)') != $PKG_VERSION ]]; then exit 1; fi  # [unix]
         - python -c "import cv2; assert 'Unknown' not in cv2.videoio_registry.getBackendName(cv2.CAP_V4L)"  # [linux]
         - python -c "import cv2, re; assert re.search('Lapack:\s+YES', cv2.getBuildInformation())"
+        - test $(pip list | grep opencv-python | tr -s " " | grep $PKG_VERSION | wc -l) -eq 1  # [unix]
```
Comment (hmaarrfk) [inline recipe/meta.yaml:257]: I personally think that the `tr` and the second `grep` commands are a little much.

Comment (traversaro) [reply recipe/meta.yaml]: Cool, it was not working (I copied it from https://github.com/conda-forge/idyntree-feedstock/blob/master/recipe/meta.yaml#L70). Let's try to remove and check if it improves.

Comment (hmaarrfk) [reply recipe/meta.yaml]: i don't think it is the issue. I would add a `pip list` above just to see the output to help debug. It should help "visually test" on windows as well.

**Comment by traversaro** [comment]
> It seems that something is being installed:
> 
> ```
> 2022-02-13T16:03:28.9066073Z -- Installing: $PREFIX/lib/python3.8/site-packages/cv2.cpython-38-x86_64-linux-gnu.so
> 2022-02-13T16:03:28.9220392Z -- Installing: $PREFIX/lib/python3.8/site-packages/opencv-python-4.5.5.dist-info/METADATA
> 2022-02-13T16:03:28.9221498Z -- Installing: $PREFIX/lib/python3.8/site-packages/opencv-python-4.5.5.dist-info/INSTALLER
> ```
> 
> I wonder what else is needed

Let's check the output of pip list, perhaps the problem is in the test.

File: recipe/meta.yaml (hunk 1 above)
Comment (traversaro) [reply recipe/meta.yaml]: I already did this in https://github.com/conda-forge/opencv-feedstock/pull/300/commits/cd3d201d06b23704dba7233835e82148ec2ad39a . : )

**Comment by traversaro** [comment]
The problem seems indeed in the version reported, that is reported as "python" instead of the actual version.
~~~
2022-02-13T17:37:44.6485710Z Package       Version
2022-02-13T17:37:44.6487111Z ------------- -------
2022-02-13T17:37:44.6489943Z numpy         1.21.5
2022-02-13T17:37:44.6496048Z opencv-python python
2022-02-13T17:37:44.6501274Z pip           22.0.3
2022-02-13T17:37:44.6506531Z setuptools    60.8.2
2022-02-13T17:37:44.6510591Z wheel         0.37.1
~~~
I checked and the same thing happens with the pip-installed metadata, if the directory is called `opencv-python-4.5.5.62.dist-info`. If the directory is called `opencv_python-4.5.5.62.dist-info` (as it happens when is installed) everything works fine.

File: recipe/install_pip_metadata.patch [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
++    set(NEW_LINE "\n")
++  endif()
++  file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/METADATA "")
++  file(APPEND ${CMAKE_CURRENT_BINARY_DIR}/METADATA "Metadata-Version: 2.1${NEW_LINE}")
++  file(APPEND ${CMAKE_CURRENT_BINARY_DIR}/METADATA "Name: opencv-python${NEW_LINE}")
++  file(APPEND ${CMAKE_CURRENT_BINARY_DIR}/METADATA "Version: ${OPENCV_VERSION}${NEW_LINE}")
++  file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/INSTALLER "${OPENCV_PYTHON_PIP_METADATA_INSTALLER}${NEW_LINE}")
++  install(
++    FILES "${CMAKE_CURRENT_BINARY_DIR}/METADATA" "${CMAKE_CURRENT_BINARY_DIR}/INSTALLER"
++    DESTINATION ${OPENCV_PYTHON_INSTALL_PATH}/opencv-python-${OPENCV_VERSION}.dist-info
```
Comment (hmaarrfk) [inline recipe/install_pip_metadata.patch]:
```suggestion
+    DESTINATION ${OPENCV_PYTHON_INSTALL_PATH}/opencv_python-${OPENCV_VERSION}.dist-info
```

From pip, there are a few things that are installed:
image

Finally, I never understood why pip had a whole installation structure
image

This was true even before they switched to ABI3 support

**Comment by hmaarrfk** [comment]
It seems we both found the problem though looking in different ways!

I think that the `tr` test is correct, it just seems "too complete".

I think that the test
```
pip list | grep opencv_python
```
is more than enough and just easier to read

**Comment by traversaro** [comment]
By the way, I think the explanation for the need for this directory name is found in https://packaging.python.org/en/latest/specifications/recording-installed-packages/ : 
> This directory is named as {name}-{version}.dist-info, with name and version fields corresponding to [Core metadata specifications](https://packaging.python.org/en/latest/specifications/core-metadata/#core-metadata). Both fields must be normalized (see [PEP 503](https://www.python.org/dev/peps/pep-0503#normalized-names) and [PEP 440](https://www.python.org/dev/peps/pep-0440#normalization) for the definition of normalization for each field respectively), and replace dash (-) characters with underscore (_) chacaters, so the .dist-info directory always has exactly one dash (-) character in its stem, separating the name and version fields.

File: recipe/meta.yaml [hunk 3]
```diff
@@ -31,7 +31,8 @@ source:
   - url: https://codeload.github.com/opencv/opencv/tar.gz/{{ version }}
     fn: opencv-{{ version }}.tar.gz
     sha256: a1cfdcf6619387ca9e232687504da996aaa9f7b5689986b8331ec02cb61d28ad
-    # patches:
+    patches:
+      - install_pip_metadata.patch
```
[the remark this answers is not in the document]
Comment (hmaarrfk) [reply recipe/meta.yaml:35]: When you do submit it, if you could cross reference it in the `meta.yaml` that would be great!

[the remark this answers is not in the document]
Comment (traversaro) [reply recipe/meta.yaml:35]: That was exactly my plan, and the reason why everything is controlled via options that by default are set to `OFF`  for backward compatibility. I just wanted to iterate on this before opening a PR to catch problems (such as the `opencv_python` vs `opencv-python` problem).

File: recipe/bld.bat [hunk 4]
```diff
@@ -103,6 +103,8 @@ cmake -LAH -G "Ninja"
     -DPYTHON3_LIBRARY=%UNIX_PREFIX%/libs/%PY_LIB%                                   ^
     -DPYTHON3_PACKAGES_PATH=%UNIX_SP_DIR%                                           ^
     -DOPENCV_PYTHON3_INSTALL_PATH=%UNIX_SP_DIR%                                     ^
+    -DOPENCV_PYTHON_PIP_METADATA_INSTALL=ON                                         ^
+    -DOPENCV_PYTHON_PIP_METADATA_INSTALLER:STRING="conda"                           ^
```
[the remark this answers is not in the document]
Comment (traversaro) [reply recipe/bld.bat:107]: Yes, this provides a helpful message if somebody tries to uninstall the package via pip, explaining that one shold use ${installer} to uninstall the package instead.

[the remark this answers is not in the document]
Comment (hmaarrfk) [reply recipe/bld.bat:107]: cool!

**Comment by hmaarrfk** [comment]
ok this seems good to go. Ping me when you get a chance to make that PR upstream then we can merge.

**Comment by hmaarrfk** [comment]
Thank you. 

I like to include a reference to the upstream issue so I added one

**Pull request merged**